### ISSUE TITLE
[RFC] Provide an API for Structural Performance Tips in DDP

### ISSUE THREAD

**Issue by SciPioneer** [pytorch/pytorch#57593]
### I. Improving Unstructured and Scattered Performance Tips

Currently a few performance-related tips/warnings w.r.t. DDP are either listed in different sources (e.g., PyTorch Lightning has a dedicated webpage for [performance tips](https://pytorch-lightning.readthedocs.io/en/stable/benchmarking/performance.html
)), or scattered in the log in the form of warnings. As a result, these valuable tips are often 1) not known by the users who have not seen these performance tips, 2) buried by a much larger volume of logs, or 3) noticed by the user but forgotten to be applied.

For example, DDP `reducer.cpp` has the following warning to suggest turning off unnecessary `find_unused_parameters` for performance gain.

```
  // Warn user about unnecessary perf hit if all parameters were used in
  // forward.
  if (unused_parameters_.empty()) {
    TORCH_WARN_ONCE(
        "find_unused_parameters=True was specified in DDP constructor, "
        "but did not find any unused parameters in the forward pass. This flag "
        "results in an extra traversal of the autograd graph every iteration, "
        " which can adversely affect performance. If your model indeed never "
        "has any unused parameters in the forward pass, consider turning this "
        "flag off. Note that this warning may be a false positive if your model "
        "has flow control causing later iterations to have unused parameters.");
  }
```

However, I have seen some users who run large-scale training and have a strong motivation to accelerate trainings are still blind to some useful performance tips like this, mainly because this warning is buried in the noisy log.

### II. Auto-sync Performance Tips and New Features

As we have been adding new features, it will be great to create a portal that can sync relevant performance tips.

For example, when a recent new feature, a DDP communication hook is applied for gradient compression, it may be also beneficial to increase the DDP bucket size to rebalance communication-computation ratio. Or when the static graph is static, we can remind the users to apply delayed AllReduce, which is a new feature released soon.

### III. Exporting Structural Performance Tips to ML Workflow Orchestration Layer

There are many ML workflow orchestration products that already cover PyTorch DDP and probably a few other training libraries/paradigms. If we can collect all the performance tips in a structural way (for DDP as a starting point), both private and OSS orchestration products can use this information to better remind users in certain ways (e.g., email, notification reminder, creating a task for improvement). These enforcement behaviors are more specific to the deployment environment. It will be nice if we can provide a common interface of a set of structural performance tips, which can be fed to the orchestration layer for deploying custom reminders. Additionally, the adoption of these reminders can give a way of tracking performance wins (and probably help quantify the impact of different features and prioritize feature enhancements/development.

The idea is very similar to democratizing a programming style (which can certainly include some performance tips). A guidance or some pure documentation won't be very effective, unless each rule can be incorporated into an automation process (e.g., a linter) that can enforce users to apply the suggestions.

## Pitch

Define a common interface (essentially a **performance tip container** of DDP) to support registering, displaying, and exporting structural performance tips. Some tips can be developed based on some internal structures of DDP, some of them may require enabling torch profiler.

1. Whenever a performance related warning is triggered in DDP, should also register it in the container, so all the tips can be collected in the same place, not scattered in the log.
2. Every performance tip object should define a `str()` for displaying.
3. All the performance tips can be exported (to the ML workflow orchestration layer) as a batch. The export method can be customized.

**Comment by rohan-varma** [comment]
Thanks for hashing out this proposal!

Regarding the `performance tip container`, I was wondering about the concrete APIs we would like to have here. It seems like the important data are:
- Why the performance improvement tip was logged (i.e. running with find_unused_parameters=True but graph may not have unused parameters)
- Reason perf is negatively impacted
- Clear action item regarding to fix/modify flags accordingly

Eventually, would we want to enhance this process by providing users the option to set a mode where we automatically enable these performance tips if we can during training? For example, the user can let DDP know whether they prefer to auto-optimize instead of having to tune flags such as `find_unused_parameters`, bucket_size, gradient comm. hook, etc and we could set these to their optimal values in one shot.

**Comment by zhaojuanmao** [comment]
I think auto tuning can help with this

**Comment by SciPioneer** [comment]
> I think auto tuning can help with this

I think these two approaches can complete each other.

First, we should do auto-tuning as much as we can, and hence minimize the actions needed from the user end.

Second, I am not sure if we can tune everything at the runtime. For example, some configs (like NCCL parameters) must be set before DDP is constructed. For some parameters, we may only know that we should tune it, but it can be hard or impractical to suggest a good value (probably because the tuning at runtime is prohibitively expensive.)

**Comment by SciPioneer** [comment]
> Thanks for hashing out this proposal!
> 
> Regarding the `performance tip container`, I was wondering about the concrete APIs we would like to have here. It seems like the important data are:
> 
> * Why the performance improvement tip was logged (i.e. running with find_unused_parameters=True but graph may not have unused parameters)
> * Reason perf is negatively impacted
> * Clear action item regarding to fix/modify flags accordingly
> 
> Eventually, would we want to enhance this process by providing users the option to set a mode where we automatically enable these performance tips if we can during training? For example, the user can let DDP know whether they prefer to auto-optimize instead of having to tune flags such as `find_unused_parameters`, bucket_size, gradient comm. hook, etc and we could set these to their optimal values in one shot.

Thanks for brainstorming the concrete API!

Regarding the follow-up questions, I think we can categorize all the configs into three types:
1) We can tune them at runtime, then no need to require any action from the user end.
2) We can tune them offline (as a centralized service in enterprise environment), and then provide a set of better configs along with a reminder. Particularly for some binary configs, we probably don't need to tune them at all. Just setting some heuristics should be enough.
3) We don't have enough resource to tune them offline, and instead we can provide instructions/tools/special modes for users to run tuning on the client side.

 I added a "Non-Goals" section to explain this.
